src/main/java/net/sf/jabref/logic/search/rules/GrammarBasedSearchRule.java

```diff
@@ -142,16 +141,25 @@ public Comparator(String field, String value, ComparisonOperator operator, boole
 
         public boolean compare(BibEntry entry) {
             // special case for searching for entrytype=phdthesis
-            if (fieldPattern.matcher("entrytype").matches()) {
+            if (fieldPattern.matcher(BibEntry.TYPE_HEADER).matches()) {
                 return matchFieldValue(entry.getType());
             }
 
+            // special case for searching a single keyword
+            if (fieldPattern.matcher("anykeyword").matches()) {
+                return entry.getKeywords().stream().anyMatch(this::matchFieldValue);
+            }
+
             // specification of fieldsKeys to search is done in the search expression itself
             Set<String> fieldsKeys = entry.getFieldNames();
 
-            List<String> matchedFieldKeys = fieldsKeys.stream().filter(matchFieldKey()).collect(Collectors.toList());
+            // special case for searching allfields=cat and title=dog
+            if (!fieldPattern.matcher("anyfield").matches()) {
+                // Filter out the requested fields
+                fieldsKeys = fieldsKeys.stream().filter(matchFieldKey()).collect(Collectors.toSet());
+            }
 
-            for (String field : matchedFieldKeys) {
+            for (String field : fieldsKeys) {
```

Comment:
You could here do the same as above, just with `return fieldsKeys.stream().filter(this::matchFieldValue).isPresent()`
The nice thing is that all operations like map., filter etc  are only performed if the value inside the optional is present, so absolutely type safe